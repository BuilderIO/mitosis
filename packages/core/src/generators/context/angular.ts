import { stringifyContextValue } from '@/helpers/get-state-object-string';
import { MitosisContext } from '@/types/mitosis-context';
import { format } from 'prettier/standalone';

type ContextToAngularOptions = {
  format?: boolean;
  typescript?: boolean;
};

export const contextToAngular =
  (options: ContextToAngularOptions = { typescript: false }) =>
  ({ context }: { context: MitosisContext }): string => {
    let str = `
import { Injectable } from '@angular/core';

@Injectable({
    providedIn: 'root'
})
export default class ${context.name}Context {
    ${stringifyContextValue(context.value)
      .replace(/^\{/, '')
      .replace(/\}$/, '')
      .replaceAll(',', ';\n')
      .replaceAll(':', ': any = ')}
    constructor() { }
}
  `;

    if (options.format !== false) {
      try {
        str = format(str, {
          parser: 'typescript',
          plugins: [
            require('prettier/parser-typescript'), // To support running in browsers
          ],
        });
      } catch (err) {
        console.error('Format error for file:', str);
        throw err;
      }
    }

    return str;
  };
