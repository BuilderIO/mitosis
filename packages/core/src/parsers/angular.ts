import ts from 'typescript';
import { parseTemplate } from '@angular/compiler';
import {
  Node,
  Element,
  Template,
  Text,
  BoundText,
} from '@angular/compiler/src/render3/r3_ast';
import { ASTWithSource } from '@angular/compiler/src/expression_parser/ast';
import { createMitosisComponent } from '../helpers/create-mitosis-component';
import { createMitosisNode } from '../helpers/create-mitosis-node';
import { MitosisNode } from '../types/mitosis-node';
import { omit } from 'lodash';
import { babelTransformCode } from '../helpers/babel-transform';
import { types } from '@babel/core';
import { MitosisComponent } from '../types/mitosis-component';
import { capitalize } from '../helpers/capitalize';

const getTsAST = (code: string) => {
  return ts.createSourceFile('code.ts', code, ts.ScriptTarget.Latest, true);
};

interface AngularToMitosisOptions {}

const transformBinding = (
  binding: string,
  _options: AngularToMitosisOptions,
) => {
  return babelTransformCode(binding, {
    Identifier(path) {
      const name = path.node.name;

      if (
        (types.isObjectProperty(path.parent) &&
          path.parent.key === path.node) ||
        (types.isMemberExpression(path.parent) &&
          path.parent.property === path.node)
      ) {
        return;
      }

      if (
        !(name.startsWith('state.') || name === 'event' || name === '$event')
      ) {
        path.replaceWith(types.identifier(`state.${name}`));
      }
    },
  });
};

const isElement = (node: Node): node is Element =>
  // TODO: theres got to be a better way than this
  Array.isArray((node as any).attributes);

const isTemplate = (node: Node): node is Template =>
  // TODO: theres got to be a better way than this
  Array.isArray((node as any).templateAttrs);

const isText = (node: Node): node is Text =>
  typeof (node as any).value === 'string';

const isBoundText = (node: Node): node is BoundText =>
  typeof (node as any).value === 'object';

const angularTemplateNodeToMitosisNode = (
  node: Node,
  options: AngularToMitosisOptions,
): MitosisNode => {
  if (isTemplate(node)) {
    const ngIf = node.templateAttrs.find((item) => item.name === 'ngIf');
    if (ngIf) {
      return createMitosisNode({
        name: 'Show',
        bindings: {
          when: {
            code: transformBinding(
              (ngIf.value as ASTWithSource).source!,
              options,
            ),
          },
        },
        children: [
          angularTemplateNodeToMitosisNode(
            omit(node, 'templateAttrs'),
            options,
          ),
        ],
      });
    }
    const ngFor = node.templateAttrs.find((item) => item.name === 'ngFor');
    if (ngFor) {
      const value = (ngFor.value as ASTWithSource).source!;
      const split = value.split(/let\s|\sof\s/);
      const [_let, itemName, _of, expression] = split;
      return createMitosisNode({
        name: 'For',
        bindings: {
          each: { code: transformBinding(expression, options) },
        },
        properties: {
          _forName: itemName,
        },
        children: [
          angularTemplateNodeToMitosisNode(
            omit(node, 'templateAttrs'),
            options,
          ),
        ],
      });
    }
  }

  if (isElement(node)) {
    const properties: Record<string, string> = {};
    const bindings: Record<string, { code: string; arguments?: string[] }> = {};

    for (const input of node.inputs) {
      bindings[input.name] = {
        code: transformBinding((input.value as ASTWithSource).source!, options),
      };
    }
    for (const output of node.outputs) {
      bindings['on' + capitalize(output.name)] = {
        code: transformBinding(
          (output.handler as ASTWithSource)
            .source! // TODO: proper reference replace
            .replace(/\$event/g, 'event'),
          options,
        ),
      };
    }
    for (const attribute of node.attributes) {
      properties[attribute.name] = attribute.value;
    }

    return createMitosisNode({
      name: node.name,
      properties,
      bindings,
      children: node.children.map((node) =>
        angularTemplateNodeToMitosisNode(node, options),
      ),
    });
  }

  if (isText(node)) {
    return createMitosisNode({
      properties: {
        _text: node.value,
      },
    });
  }

  if (isBoundText(node)) {
    // TODO: handle the bindings
    return createMitosisNode({
      properties: {
        _text: (node.value as ASTWithSource).source!,
      },
    });
  }

  throw new Error(`Element node type {${node}} is not supported`);
};

const angularTemplateToMitosisNodes = (
  template: string,
  options: AngularToMitosisOptions,
) => {
  const ast = parseTemplate(template, '.');
  const blocks = ast.nodes.map((node) =>
    angularTemplateNodeToMitosisNode(node, options),
  );

  return blocks;
};

const parseTypescript = (code: string, options: AngularToMitosisOptions) => {
  const component = createMitosisComponent();

  const ast = getTsAST(code);
  for (const statement of ast.statements) {
    if (ts.isClassDeclaration(statement)) {
      if (statement.decorators) {
        for (const decorator of statement.decorators) {
          // TODO: proper reference tracing
          if (ts.isCallExpression(decorator.expression))
            if (
              ts.isIdentifier(decorator.expression.expression) &&
              decorator.expression.expression.text === 'Component'
            ) {
              const firstArg = decorator.expression.arguments[0];
              if (ts.isObjectLiteralExpression(firstArg)) {
                firstArg.properties.find((item) => {
                  if (ts.isPropertyAssignment(item)) {
                    if (
                      ts.isIdentifier(item.name) &&
                      item.name.text === 'template'
                    ) {
                      if (ts.isTemplateLiteral(item.initializer)) {
                        const template = item.initializer
                          .getText()
                          .trim()
                          .slice(1, -1);
                        component.children = angularTemplateToMitosisNodes(
                          template,
                          options,
                        );
                      }
                    }
                  }
                });
              }
            }
        }
      }
    }
  }

  return component;
};

export function angularToMitosisComponent(
  code: string,
  options: AngularToMitosisOptions = {},
): MitosisComponent {
  return parseTypescript(code, options);
}
