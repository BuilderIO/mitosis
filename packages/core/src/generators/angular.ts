import dedent from 'dedent';
import { format } from 'prettier/standalone';
import { collectCss } from '../helpers/collect-styles';
import { fastClone } from '../helpers/fast-clone';
import { getRefs } from '../helpers/get-refs';
import { getStateObjectStringFromComponent } from '../helpers/get-state-object-string';
import { mapRefs } from '../helpers/map-refs';
import { renderPreComponent } from '../helpers/render-imports';
import { stripStateAndPropsRefs } from '../helpers/strip-state-and-props-refs';
import { selfClosingTags } from '../parsers/jsx';
import { MitosisNode } from '../types/mitosis-node';
import {
  runPostCodePlugins,
  runPostJsonPlugins,
  runPreCodePlugins,
  runPreJsonPlugins,
} from '../modules/plugins';
import isChildren from '../helpers/is-children';
import { getProps } from '../helpers/get-props';
import { kebabCase } from 'lodash';
import { stripMetaProperties } from '../helpers/strip-meta-properties';
import { removeSurroundingBlock } from '../helpers/remove-surrounding-block';
import { BaseTranspilerOptions, Transpiler } from '../types/config';
import { indent } from '../helpers/indent';

export interface ToAngularOptions extends BaseTranspilerOptions {}

const mappers: {
  [key: string]: (json: MitosisNode, options: ToAngularOptions) => string;
} = {
  Fragment: (json, options) => {
    return `<div>${json.children
      .map((item) => blockToAngular(item, options))
      .join('\n')}</div>`;
  },
};

export const blockToAngular = (
  json: MitosisNode,
  options: ToAngularOptions = {},
): string => {
  if (mappers[json.name]) {
    return mappers[json.name](json, options);
  }

  if (isChildren(json)) {
    return `<ng-content></ng-content>`;
  }

  if (json.properties._text) {
    return json.properties._text;
  }

  if (json.bindings._text) {
    return `{{${stripStateAndPropsRefs(json.bindings._text as string)}}}`;
  }

  let str = '';

  if (json.name === 'For') {
    str += `<ng-container *ngFor="let ${
      json.properties._forName
    } of ${stripStateAndPropsRefs(json.bindings.each as string)}">`;
    str += json.children
      .map((item) => blockToAngular(item, options))
      .join('\n');
    str += `</ng-container>`;
  } else if (json.name === 'Show') {
    str += `<ng-container *ngIf="${stripStateAndPropsRefs(
      json.bindings.when as string,
    )}">`;
    str += json.children
      .map((item) => blockToAngular(item, options))
      .join('\n');
    str += `</ng-container>`;
  } else {
    str += `<${json.name} `;

    // TODO: spread support for angular
    // if (json.bindings._spread) {
    //   str += `v-bind="${stripStateAndPropsRefs(
    //     json.bindings._spread as string,
    //   )}"`;
    // }

    for (const key in json.properties) {
      if (key.startsWith('$')) {
        continue;
      }
      const value = json.properties[key];
      str += ` ${key}="${value}" `;
    }
    for (const key in json.bindings) {
      if (key === '_spread') {
        continue;
      }
      if (key.startsWith('$')) {
        continue;
      }
      const value = json.bindings[key] as string;
      // TODO: proper babel transform to replace. Util for this
      const useValue = stripStateAndPropsRefs(value);

      if (key.startsWith('on')) {
        let event = key.replace('on', '').toLowerCase();
        if (
          event === 'change' &&
          json.name === 'input' /* todo: other tags */
        ) {
          event = 'input';
        }
        // TODO: proper babel transform to replace. Util for this
        const finalValue = removeSurroundingBlock(
          useValue.replace(/event\./g, '$event.'),
        );
        str += ` (${event})="${finalValue}" `;
      } else if (key === 'ref') {
        str += ` #${useValue} `;
      } else {
        str += ` [${key}]="${useValue}" `;
      }
    }
    if (selfClosingTags.has(json.name)) {
      return str + ' />';
    }
    str += '>';
    if (json.children) {
      str += json.children
        .map((item) => blockToAngular(item, options))
        .join('\n');
    }

    str += `</${json.name}>`;
  }
  return str;
};

export const componentToAngular =
  (options: ToAngularOptions = {}): Transpiler =>
  ({ component }) => {
    // Make a copy we can safely mutate, similar to babel's toolchain
    let json = fastClone(component);
    if (options.plugins) {
      json = runPreJsonPlugins(json, options.plugins);
    }

    const props = getProps(component);

    const refs = Array.from(getRefs(json));
    mapRefs(json, (refName) => `this.${refName}.nativeElement`);

    if (options.plugins) {
      json = runPostJsonPlugins(json, options.plugins);
    }
    let css = collectCss(json);
    if (options.prettier !== false) {
      css = tryFormat(css, 'css');
    }

    let template = json.children.map((item) => blockToAngular(item)).join('\n');
    if (options.prettier !== false) {
      template = tryFormat(template, 'html');
    }

    stripMetaProperties(json);

    const dataString = getStateObjectStringFromComponent(json, {
      format: 'class',
      valueMapper: (code) =>
        stripStateAndPropsRefs(code, { replaceWith: 'this.' }),
    });

    let str = dedent`
    import { Component ${refs.length ? ', ViewChild, ElementRef' : ''}${
      props.size ? ', Input' : ''
    } } from '@angular/core';
    ${renderPreComponent(json)}

    @Component({
      selector: '${kebabCase(json.name || 'my-component')}',
      template: \`
        ${indent(template, 8).replace(/`/g, '\\`').replace(/\$\{/g, '\\${')}
      \`,
      ${
        css.length
          ? `styles: [
        \`${indent(css, 8)}\`
      ],`
          : ''
      }
    })
    export default class ${component.name} {
      ${Array.from(props)
        .map((item) => `@Input() ${item}: any`)
        .join('\n')}

      ${refs
        .map((refName) => `@ViewChild('${refName}') ${refName}: ElementRef`)
        .join('\n')}

      ${
        !component.hooks.onMount
          ? ''
          : `ngOnInit() {
              ${stripStateAndPropsRefs(component.hooks.onMount.code, {
                replaceWith: 'this.',
              })}
            }`
      }

      ${
        !component.hooks.onUpdate?.length
          ? ''
          : `ngAfterContentChecked() {
              ${component.hooks.onUpdate.map((hook) =>
                stripStateAndPropsRefs(hook.code, {
                  replaceWith: 'this.',
                }),
              )}
            }`
      }

      ${
        !component.hooks.onUnMount
          ? ''
          : `ngOnDestroy() {
              ${stripStateAndPropsRefs(component.hooks.onUnMount.code, {
                replaceWith: 'this.',
              })}
            }`
      }

      ${dataString}
    }
  `;

    if (options.plugins) {
      str = runPreCodePlugins(str, options.plugins);
    }
    if (options.prettier !== false) {
      str = tryFormat(str, 'typescript');
    }
    if (options.plugins) {
      str = runPostCodePlugins(str, options.plugins);
    }
    return str;
  };

const tryFormat = (str: string, parser: string) => {
  try {
    return format(str, {
      parser,
      plugins: [
        // To support running in browsers
        require('prettier/parser-typescript'),
        require('prettier/parser-postcss'),
        require('prettier/parser-html'),
        require('prettier/parser-babel'),
      ],
      htmlWhitespaceSensitivity: 'ignore',
    });
  } catch (err) {
    console.warn('Could not prettify', { string: str }, err);
  }
  return str;
};
