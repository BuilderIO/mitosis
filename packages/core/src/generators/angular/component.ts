import { blockToAngular } from '@/generators/angular/blocks';
import {
  HELPER_FUNCTIONS,
  generateNgModule,
  getAppropriateTemplateFunctionKeys,
  getDefaultProps,
  preprocessCssAsJson,
  processAngularCode,
  transformState,
  traverseAndCheckIfInnerHTMLIsUsed,
  traverseToGetAllDynamicComponents,
} from '@/generators/angular/helpers';
import { getInputImports, getInputs } from '@/generators/angular/helpers/get-inputs';
import { getOutputImports, getOutputs } from '@/generators/angular/helpers/get-outputs';
import { addCodeNgAfterViewInit } from '@/generators/angular/helpers/hooks';
import { getClassPropertiesPlugin } from '@/generators/angular/plugins/get-class-properties-plugin';
import { getCodeProcessorPlugins } from '@/generators/angular/plugins/get-code-processor-plugins';
import {
  BUILT_IN_COMPONENTS,
  DEFAULT_ANGULAR_OPTIONS,
  ToAngularOptions,
} from '@/generators/angular/types';
import { stringifySingleScopeOnMount } from '@/generators/helpers/on-mount';
import { dashCase } from '@/helpers/dash-case';
import { dedent } from '@/helpers/dedent';
import { fastClone } from '@/helpers/fast-clone';
import { getChildComponents } from '@/helpers/get-child-components';
import { getComponentsUsed } from '@/helpers/get-components-used';
import { getCustomImports } from '@/helpers/get-custom-imports';
import { getPropFunctions } from '@/helpers/get-prop-functions';
import { getProps } from '@/helpers/get-props';
import { getPropsRef } from '@/helpers/get-props-ref';
import { getRefs } from '@/helpers/get-refs';
import { getStateObjectStringFromComponent } from '@/helpers/get-state-object-string';
import { getTypedFunction } from '@/helpers/get-typed-function';
import { indent } from '@/helpers/indent';
import { isUpperCase } from '@/helpers/is-upper-case';
import { mapRefs } from '@/helpers/map-refs';
import { initializeOptions } from '@/helpers/merge-options';
import { renderPreComponent } from '@/helpers/render-imports';
import { stripMetaProperties } from '@/helpers/strip-meta-properties';
import { collectCss } from '@/helpers/styles/collect-css';
import {
  ROOT_REF,
  getAddAttributePassingRef,
  getAttributePassingString,
  shouldAddAttributePassing,
} from '@/helpers/web-components/attribute-passing';
import {
  runPostCodePlugins,
  runPostJsonPlugins,
  runPreCodePlugins,
  runPreJsonPlugins,
} from '@/modules/plugins';
import { TranspilerGenerator } from '@/types/transpiler';
import { kebabCase, uniq } from 'lodash';
import { tryFormat } from './helpers/format';

export const componentToAngular: TranspilerGenerator<ToAngularOptions> = (userOptions = {}) => {
  return ({ component: _component }) => {
    // Make a copy we can safely mutate, similar to babel's toolchain
    let json = fastClone(_component);

    const useMetadata = json.meta?.useMetadata;

    const contextVars: string[] = Object.keys(json?.context?.get || {});
    const metaOutputVars: string[] = (useMetadata?.outputs as string[]) || [];

    const outputVars: string[] = uniq([...metaOutputVars, ...getPropFunctions(json)]);

    const options = initializeOptions({
      target: 'angular',
      component: _component,
      defaults: DEFAULT_ANGULAR_OPTIONS,
      userOptions,
    });
    options.plugins = getCodeProcessorPlugins({ json, options, outputVars, contextVars });

    if (options.state === 'class-properties') {
      options.plugins.push(getClassPropertiesPlugin());
    }

    if (options.plugins) {
      json = runPreJsonPlugins({ json, plugins: options.plugins });
    }

    // Prepare Props
    const [forwardProp, hasPropRef] = getPropsRef(json, true);
    const props = getProps(json);
    // prevent jsx props from showing up as @Input
    if (hasPropRef) {
      props.delete(forwardProp);
    }
    props.delete('children');

    // remove props for outputs
    for (const variableName of outputVars) {
      props.delete(variableName);
    }

    const childComponents: string[] = getChildComponents(json);

    const customImports: string[] = getCustomImports(json);

    const { exports: localExports = {} } = json;
    const localExportVars: string[] = Object.keys(localExports)
      .filter((key) => localExports[key].usedInLocal)
      .map((key) => `${key} = ${key};`);

    // Context handling
    const injectables: string[] = contextVars.map((variableName) => {
      const variableType = json?.context?.get[variableName].name;
      if (options?.experimental?.injectables) {
        return options?.experimental?.injectables(variableName, variableType);
      }
      if (options?.experimental?.inject) {
        return `@Inject(forwardRef(() => ${variableType})) public ${variableName}: ${variableType}`;
      }
      return `public ${variableName} : ${variableType}`;
    });

    // Handle refs
    const domRefs = getRefs(json);
    const jsRefs = Object.keys(json.refs).filter((ref) => !domRefs.has(ref));

    const withAttributePassing = shouldAddAttributePassing(json, options);
    const rootRef = getAddAttributePassingRef(json, options);
    if (withAttributePassing) {
      if (!domRefs.has(rootRef)) {
        domRefs.add(rootRef);
      }

      addCodeNgAfterViewInit(
        json,
        `
            const element: HTMLElement | null = this.${rootRef}?.nativeElement;
            this.enableAttributePassing(element, "${dashCase(json.name)}");
            `,
      );
    }

    mapRefs(json, (refName) => {
      const isDomRef = domRefs.has(refName);
      return `this.${isDomRef ? '' : '_'}${refName}${isDomRef ? '.nativeElement' : ''}`;
    });

    if (options.plugins) {
      json = runPostJsonPlugins({ json, plugins: options.plugins });
    }

    // CSS
    preprocessCssAsJson(json);
    let css = collectCss(json);
    if (options.prettier !== false) {
      css = tryFormat(css, 'css');
    }

    const hostDisplayCss = options.visuallyIgnoreHostElement ? ':host { display: contents; }' : '';
    const styles = css.length ? [hostDisplayCss, css].join('\n') : hostDisplayCss;

    const helperFunctions = new Set<string>();
    const shouldUseSanitizer =
      !useMetadata?.angular?.sanitizeInnerHTML && traverseAndCheckIfInnerHTMLIsUsed(json);
    const changeDetectionStrategy = useMetadata?.angular?.changeDetection;

    let template = json.children
      .map((item) => {
        const tmpl = blockToAngular({
          root: json,
          json: item,
          options,
          rootRef: withAttributePassing && rootRef === ROOT_REF ? rootRef : undefined, // only pass rootRef if it's not the default
          blockOptions: {
            childComponents,
            nativeAttributes: useMetadata?.angular?.nativeAttributes ?? [],
            nativeEvents: useMetadata?.angular?.nativeEvents ?? [],
            sanitizeInnerHTML: !shouldUseSanitizer,
          },
        });
        if (options.state === 'inline-with-wrappers') {
          getAppropriateTemplateFunctionKeys(tmpl).forEach((key) =>
            helperFunctions.add(HELPER_FUNCTIONS(options.typescript)[key]),
          );
        }
        return tmpl;
      })
      .join('\n');

    if (options.prettier !== false) {
      template = tryFormat(template, 'html');
    }

    stripMetaProperties(json);

    const { components: dynamicComponents, dynamicTemplate } = traverseToGetAllDynamicComponents(
      json,
      options,
      {
        childComponents,
        nativeAttributes: useMetadata?.angular?.nativeAttributes ?? [],
        nativeEvents: useMetadata?.angular?.nativeEvents ?? [],
      },
    );

    transformState(json);

    const dataString = getStateObjectStringFromComponent(json, {
      format: 'class',
      withType: options.typescript,
      valueMapper: (code, type, typeParameter) => {
        let value = code;
        if (type !== 'data') {
          value = getTypedFunction(code, options.typescript, typeParameter);
        }

        if (options.api === 'signals') {
          if (type === 'data') {
            value = `signal(${value})`;
          }
        }

        return processAngularCode({
          replaceWith: 'this',
          contextVars,
          outputVars,
          domRefs: Array.from(domRefs),
        })(value);
      },
    });

    const refsForObjSpread = getRefs(json, 'spreadRef');

    // Preparing built in component metadata parameters
    const componentsUsed = Array.from(getComponentsUsed(json)).filter(
      (item) => item.length && isUpperCase(item[0]) && !BUILT_IN_COMPONENTS.has(item),
    );
    const componentMetadata: Record<string, any> = {
      selector: useMetadata?.angular?.selector
        ? `'${useMetadata?.angular?.selector}'`
        : `'${kebabCase(json.name || 'my-component')}'`,
      template: `\`
          ${indent(dynamicTemplate, 8).replace(/`/g, '\\`').replace(/\$\{/g, '\\${')}
          ${indent(template, 8).replace(/`/g, '\\`').replace(/\$\{/g, '\\${')}
          \``,
      ...(changeDetectionStrategy === 'OnPush'
        ? {
            changeDetection: 'ChangeDetectionStrategy.OnPush',
          }
        : {}),
      ...(styles
        ? {
            styles: `[\`${indent(styles, 8)}\`]`,
          }
        : {}),
      ...(options.standalone
        ? // TODO: also add child component imports here as well
          {
            standalone: 'true',
            imports: `[${['CommonModule', ...componentsUsed].join(', ')}]`,
          }
        : {}),
    };
    // Taking into consideration what user has passed in options and allowing them to override the default generated metadata
    Object.entries(json.meta.angularConfig || {}).forEach(([key, value]) => {
      componentMetadata[key] = value;
    });

    const hasConstructor =
      Boolean(injectables.length) ||
      dynamicComponents.size ||
      refsForObjSpread.size ||
      shouldUseSanitizer;

    const angularCoreImports = [
      ...(outputVars.length ? getOutputImports(options.api) : []),
      ...(options?.experimental?.inject ? ['Inject', 'forwardRef'] : []),
      'Component',
      ...(domRefs.size || dynamicComponents.size || refsForObjSpread.size
        ? ['ViewChild', 'ElementRef']
        : []),
      ...(refsForObjSpread.size ? ['Renderer2'] : []),
      ...(props.size ? getInputImports(options.api) : []),
      ...(dynamicComponents.size ? ['ViewContainerRef', 'TemplateRef'] : []),
      ...(json.hooks.onUpdate?.length && options.typescript ? ['SimpleChanges'] : []),
      ...(changeDetectionStrategy === 'OnPush' ? ['ChangeDetectionStrategy'] : []),
      ...(options.api === 'signals' ? ['signal'] : []),
    ].join(', ');

    const constructorInjectables = [
      ...injectables,
      ...(dynamicComponents.size
        ? [`private vcRef${options.typescript ? ': ViewContainerRef' : ''}`]
        : []),
      ...(refsForObjSpread.size
        ? [`private renderer${options.typescript ? ': Renderer2' : ''}`]
        : []),
      ...(shouldUseSanitizer
        ? [`protected sanitizer${options.typescript ? ': DomSanitizer' : ''}`]
        : []),
    ].join(',\n');

    let str = dedent`
        import { ${angularCoreImports} } from '@angular/core';
        ${shouldUseSanitizer ? `import { DomSanitizer } from '@angular/platform-browser';` : ''}
        ${options.standalone ? `import { CommonModule } from '@angular/common';` : ''}
    
        ${json.types ? json.types.join('\n') : ''}
        ${getDefaultProps(json)}
        ${renderPreComponent({
          explicitImportFileExtension: options.explicitImportFileExtension,
          component: json,
          target: 'angular',
          excludeMitosisComponents: !options.standalone && !options.preserveImports,
          preserveFileExtensions: options.preserveFileExtensions,
          componentsUsed,
          importMapper: options?.importMapper,
        })}
    
        @Component({
          ${Object.entries(componentMetadata)
            .map(([k, v]) => `${k}: ${v}`)
            .join(',')}
        })
        export default class ${json.name} {
          ${localExportVars.join('\n')}
          ${customImports.map((name) => `${name} = ${name}`).join('\n')}
    
          ${getInputs({
            json,
            options,
            props: Array.from(props),
          })}    
          ${getOutputs({ json, outputVars, api: options.api })}
    
          ${[...Array.from(domRefs), ...Array.from(refsForObjSpread)]
            .map(
              (refName) =>
                `@ViewChild('${refName}') ${refName}${options.typescript ? '!: ElementRef' : ''}`,
            )
            .join('\n')}
    
          ${Array.from(dynamicComponents)
            .map(
              (component) =>
                `@ViewChild('${component
                  .split('.')[1]
                  .toLowerCase()}Template', { static: true }) ${component
                  .split('.')[1]
                  .toLowerCase()}TemplateRef${options.typescript ? '!: TemplateRef<any>' : ''}`,
            )
            .join('\n')}
    
          ${dynamicComponents.size ? `myContent${options.typescript ? '?: any[][];' : ''}` : ''}
          ${
            refsForObjSpread.size
              ? `_listenerFns = new Map${options.typescript ? '<string, () => void>' : ''}()`
              : ''
          }
    
          ${dataString}
    
          ${helperFunctions.size ? Array.from(helperFunctions).join('\n') : ''}
    
          ${jsRefs
            .map((ref) => {
              const argument = json.refs[ref].argument;
              const typeParameter = json.refs[ref].typeParameter;
              return `private _${ref}${typeParameter ? `: ${typeParameter}` : ''}${
                argument
                  ? ` = ${processAngularCode({
                      replaceWith: 'this.',
                      contextVars,
                      outputVars,
                      domRefs: Array.from(domRefs),
                    })(argument)}`
                  : ''
              };`;
            })
            .join('\n')}
    
          ${!hasConstructor ? '' : `constructor(\n${constructorInjectables}) {}`}
          
          ${withAttributePassing ? getAttributePassingString(options.typescript) : ''}
          
          ${
            !json.hooks.onMount.length && !dynamicComponents.size && !json.hooks.onInit?.code
              ? ''
              : `ngOnInit() {
                  ${
                    !json.hooks?.onInit
                      ? ''
                      : `
                        ${json.hooks.onInit?.code}
                        `
                  }
                  ${
                    json.hooks.onMount.length > 0
                      ? `
                        if (typeof window !== 'undefined') {
                          ${stringifySingleScopeOnMount(json)}
                        }
                        `
                      : ''
                  }
                  ${
                    dynamicComponents.size
                      ? `
                  this.myContent = [${Array.from(dynamicComponents)
                    .map(
                      (component) =>
                        `this.vcRef.createEmbeddedView(this.${component
                          .split('.')[1]
                          .toLowerCase()}TemplateRef).rootNodes`,
                    )
                    .join(', ')}];
                  `
                      : ''
                  }
                }`
          }
    
          ${
            // hooks specific to Angular
            json.compileContext?.angular?.hooks
              ? Object.entries(json.compileContext?.angular?.hooks)
                  .map(([key, value]) => {
                    return `${key}() {
                ${value.code}
              }`;
                  })
                  .join('\n')
              : ''
          }
    
          ${
            !json.hooks.onUpdate?.length
              ? ''
              : `ngOnChanges(changes${options.typescript ? ': SimpleChanges' : ''}) {
                  if (typeof window !== 'undefined') {
                    ${json.hooks.onUpdate?.reduce((code, hook) => {
                      code += hook.code;
                      return code + '\n';
                    }, '')}
                  }
                }
                    `
          }
    
          ${
            !json.hooks.onUnMount && !refsForObjSpread.size
              ? ''
              : `ngOnDestroy() {
                  ${json.hooks.onUnMount?.code || ''}
                  ${
                    refsForObjSpread.size
                      ? `for (const fn of this._listenerFns.values()) { fn(); }`
                      : ''
                  }
                }`
          }
    
        }
      `;

    if (options.standalone !== true) {
      str = generateNgModule(str, json.name, componentsUsed, json, options.bootstrapMapper);
    }
    if (options.plugins) {
      str = runPreCodePlugins({ json, code: str, plugins: options.plugins });
    }
    if (options.prettier !== false) {
      str = tryFormat(str, 'typescript');
    }
    if (options.plugins) {
      str = runPostCodePlugins({ json, code: str, plugins: options.plugins });
    }

    return str;
  };
};
