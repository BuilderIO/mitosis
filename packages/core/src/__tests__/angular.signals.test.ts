import { componentToAngular } from '@/generators/angular';
import { runTestsForTarget } from './test-generator';

describe('Angular signals', () => {
  runTestsForTarget({
    options: {
      standalone: true,
      api: 'signals',
    },
    target: 'angular',
    generator: componentToAngular,
  });
});

/*describe('Angular signals local', () => {
  runTestsForTarget({
    options: {
      standalone: true,
      api: 'signals',
      typescript: true,
    },
    target: 'angular',
    generator: componentToAngular,
    only: [
      // 'signals'
      // 'eventProps',
      // 'basicRefAttributePassing',
      // 'eventInputAndChange',
      // "spreadAttrs", // TODO: spreads
      // "dynamicComponentWithEventArg" // TODO: add dynamic components
      'Basic',
    ],
    logOutput: true,
  });
});*/
