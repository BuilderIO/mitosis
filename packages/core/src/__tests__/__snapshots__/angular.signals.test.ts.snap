// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Angular local > jsx > Javascript Test > signals 1`] = `
"import { output, Component, input } from \\"@angular/core\\";

import { CommonModule } from \\"@angular/common\\";

@Component({
  selector: \\"signals-test-component\\",
  template: \`
    <button (click)=\\"handleOutput()\\">{{testInput}}</button>
  \`,
  styles: [
    \`
      :host {
        display: contents;
      }
    \`,
  ],
  standalone: true,
  imports: [CommonModule],
})
export default class SignalsTestComponent {
  testInput = input<any>();
  testOutput = output<any>();

  handleOutput() {
    if (this.testOutput) {
      this.testOutput.emit();
    }
  }
}
"
`;

exports[`Angular local > jsx > Typescript Test > signals 1`] = `
"import { output, Component, input } from \\"@angular/core\\";

import { CommonModule } from \\"@angular/common\\";

type Props = {
  testInput: string;
  testOutput?: () => void;
};
type Store = {
  handleOutput: () => void;
};

@Component({
  selector: \\"signals-test-component\\",
  template: \`
    <button (click)=\\"handleOutput()\\">{{testInput}}</button>
  \`,
  styles: [
    \`
      :host {
        display: contents;
      }
    \`,
  ],
  standalone: true,
  imports: [CommonModule],
})
export default class SignalsTestComponent {
  testInput = input<Props[\\"testInput\\"]>();
  testOutput = output<any>();

  handleOutput(): ReturnType<Store[\\"handleOutput\\"]> {
    if (this.testOutput) {
      this.testOutput.emit();
    }
  }
}
"
`;

exports[`Angular local > svelte > Javascript Test > basic 1`] = `
"import { Component } from \\"@angular/core\\";

import { CommonModule } from \\"@angular/common\\";

@Component({
  selector: \\"my-component\\",
  template: \`
    <div>
      <input (change)=\\"name = $event.target.value\\" [attr.value]=\\"name\\" />

      Hello! I can run in React, Vue, Solid, or Liquid!
    </div>
  \`,
  styles: [
    \`
      :host {
        display: contents;
      }
    \`,
  ],
  standalone: true,
  imports: [CommonModule],
})
export default class MyComponent {
  name = \\"Steve\\";
}
"
`;

exports[`Angular local > svelte > Typescript Test > basic 1`] = `
"import { Component } from \\"@angular/core\\";

import { CommonModule } from \\"@angular/common\\";

@Component({
  selector: \\"my-component\\",
  template: \`
    <div>
      <input (change)=\\"name = $event.target.value\\" [attr.value]=\\"name\\" />

      Hello! I can run in React, Vue, Solid, or Liquid!
    </div>
  \`,
  styles: [
    \`
      :host {
        display: contents;
      }
    \`,
  ],
  standalone: true,
  imports: [CommonModule],
})
export default class MyComponent {
  name = \\"Steve\\";
}
"
`;
